lib/improver/tests/nbhood/circular_kernel/test_CircularPercentiles.py

```diff
@@ -0,0 +1,396 @@
+
+    """Test neighbourhood circular percentile plugin."""
+
+    def test_basic(self):
+        """Test that the plugin returns an iris.cube.Cube."""
+        cube = set_up_cube(
+            zero_point_indices=((0, 0, 2, 2),), num_time_points=1,
+            num_grid_points=5)
+        radius = 4000.
+        result = CircularPercentiles().run(cube, radius)
+        self.assertIsInstance(result, Cube)
+
+    def test_single_point(self):
+        """Test behaviour for a single non-zero grid cell."""
+        cube = set_up_cube()
+        expected = np.ones([15] + list(np.shape(cube.data)))
+        expected = np.transpose(expected, [1, 0, 2, 3, 4])
+        expected[0, :, 0, 7, 5] = PERCENTILES_1_IN_13
+        expected[0, :, 0, 6:9, 6] = (
+            np.transpose(np.tile(PERCENTILES_1_IN_13, (3, 1))))
+        expected[0, :, 0, 5:10, 7] = (
+            np.transpose(np.tile(PERCENTILES_1_IN_13, (5, 1))))
+        expected[0, :, 0, 6:9, 8] = (
+            np.transpose(np.tile(PERCENTILES_1_IN_13, (3, 1))))
+        expected[0, :, 0, 7, 9] = PERCENTILES_1_IN_13
+        radius = 4000.
+        result = CircularPercentiles().run(cube, radius)
+        self.assertArrayAlmostEqual(result.data, expected)
+
+    def test_multi_point_multitimes(self):
+        """Test behaviour for points over multiple times."""
+        cube = set_up_cube(
+            zero_point_indices=[(0, 0, 10, 10), (0, 1, 7, 7)],
+            num_time_points=2
+        )
+        expected = np.ones([15] + list(np.shape(cube.data)))
+        expected = np.transpose(expected, [1, 0, 2, 3, 4])
+        expected[0, :, 0, 10, 8] = PERCENTILES_1_IN_13
+        expected[0, :, 0, 9:12, 9] = (
+            np.transpose(np.tile(PERCENTILES_1_IN_13, (3, 1))))
+        expected[0, :, 0, 8:13, 10] = (
+            np.transpose(np.tile(PERCENTILES_1_IN_13, (5, 1))))
+        expected[0, :, 0, 9:12, 11] = (
+            np.transpose(np.tile(PERCENTILES_1_IN_13, (3, 1))))
+        expected[0, :, 0, 10, 12] = PERCENTILES_1_IN_13
+        expected[0, :, 1, 7, 5] = PERCENTILES_1_IN_13
+        expected[0, :, 1, 6:9, 6] = (
+            np.transpose(np.tile(PERCENTILES_1_IN_13, (3, 1))))
+        expected[0, :, 1, 5:10, 7] = (
+            np.transpose(np.tile(PERCENTILES_1_IN_13, (5, 1))))
+        expected[0, :, 1, 6:9, 8] = (
+            np.transpose(np.tile(PERCENTILES_1_IN_13, (3, 1))))
+        expected[0, :, 1, 7, 9] = PERCENTILES_1_IN_13
+        radius = 4000.
+        result = (CircularPercentiles().run(cube, radius))
+        self.assertArrayAlmostEqual(result.data, expected)
+
+    def test_single_point_lat_long(self):
+        """Test behaviour for a single grid cell on lat long grid."""
+        cube = set_up_cube_lat_long()
+        msg = "Invalid grid: projection_x/y coords required"
+        radius = 6000.
+        with self.assertRaisesRegexp(ValueError, msg):
+            CircularPercentiles().run(cube, radius)
+
+    def test_single_point_masked_to_null(self):
```

Comment:
This is masking a point with valid data in, not a point with NaNs in. The code does something quite different if there is a NaN in the place you masked, it actually spreads the NaN over the the circular neighbourhood. I think we need to think about this, as this is not the required behaviour, although this could be work for a follow on ticket.